4.5
(1)求样本相关系数矩阵R

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats

data4_5 = pd.DataFrame({
    "省(区、市)": [
        "山西", "内蒙古", "吉林", "黑龙江", "河南", "甘肃", "青海", "河北",
        "陕西", "宁夏", "新疆", "湖北", "云南", "湖南",
        "安徽", "贵州", "辽宁", "四川", "山东", "江西", "福建", "广西",
        "海南", "天津", "江苏", "浙江", "北京", "西藏", "上海", "广东"
    ],
    "x1": [
        8.35, 9.25, 8.19, 7.73, 9.42, 9.16, 10.06, 9.09,
        9.41, 8.70, 6.93, 8.67, 9.98, 6.77,
        8.14, 7.67, 7.90, 7.18, 8.82, 6.25, 10.60, 7.27,
        13.45, 10.85, 7.21, 7.68, 7.78, 7.94, 8.28, 12.47
    ],
    "x2": [
        23.53, 23.75, 30.50, 29.20, 27.93, 27.98, 28.64, 28.12,
        28.20, 28.12, 29.85, 36.05, 37.69, 38.69,
        37.75, 35.71, 39.77, 40.91, 33.70, 35.02, 52.41, 52.65,
        55.85, 44.68, 45.79, 50.37, 48.44, 39.65, 64.34, 76.39
    ],
    "x3": [
        7.51, 6.61, 4.72, 5.42, 8.20, 9.01, 10.52, 7.40,
        5.77, 7.21, 4.54, 7.31, 7.01, 6.01,
        9.61, 8.04, 8.49, 7.32, 7.59, 4.72, 7.70, 3.84,
        5.50, 7.32, 7.66, 11.35, 8.00, 20.97, 8.00, 5.52
    ],
    "x4": [
        8.62, 9.19, 9.78, 9.43, 8.14, 9.32, 10.05, 9.62,
        10.80, 10.53, 9.49, 7.75, 8.94, 8.82,
        8.49, 8.31, 12.94, 8.94, 10.98, 6.28, 9.98, 9.16,
        7.45, 14.51, 10.36, 13.30, 20.51, 20.82, 22.22, 11.24
    ],
    "x5": [
        17.42, 17.77, 16.28, 19.29, 16.17, 15.99, 16.18, 17.26,
        16.36, 19.45, 16.62, 16.67, 16.15, 14.79,
        13.15, 15.13, 19.27, 17.60, 18.82, 10.03, 12.53, 13.03,
        9.55, 17.13, 16.56, 19.25, 22.12, 22.52, 20.06, 14.52
    ],
    "x6": [
        10.00, 10.48, 7.60, 8.49, 9.42, 9.10, 8.39, 11.12,
        11.56, 13.30, 10.65, 11.68, 11.08, 11.44,
        9.76, 7.76, 11.05, 12.75, 14.73, 7.15, 11.70, 15.26,
        9.52, 12.08, 12.86, 14.59, 15.73, 12.41, 15.12, 22.00
    ],
    "x7": [
        1.04, 1.72, 2.52, 2.52, 1.55, 1.82, 1.96, 2.49,
        1.53, 1.66, 1.88, 2.38, 0.83, 1.74,
        1.28, 1.41, 2.04, 1.14, 1.78, 1.93, 2.31, 1.98,
        2.21, 1.26, 2.25, 2.75, 1.15, 1.75, 0.72, 5.46
    ],
    "x8": [
        11.21, 10.51, 10.32, 10.00, 9.76, 11.35, 10.81, 12.65,
        12.17, 11.96, 13.61, 12.88, 11.67, 13.23,
        11.28, 13.25, 13.29, 14.80, 10.10, 10.39, 14.69, 14.57,
        16.30, 11.57, 11.69, 14.87, 16.61,  7.90, 22.89, 25.50
    ]
})

In [2]:
# 计算相关系数矩阵
correlation_matrix = data4_5.select_dtypes(include=[np.number]).corr()
print("相关系数矩阵R：", correlation_matrix)

相关系数矩阵R：           x1        x2        x3        x4        x5        x6        x7  \
x1  1.000000  0.333647 -0.054539 -0.061254 -0.289361  0.198796  0.348698   
x2  0.333647  1.000000 -0.022902  0.398931 -0.156304  0.711134  0.413595   
x3 -0.054539 -0.022902  1.000000  0.533329  0.496763  0.032830 -0.139086   
x4 -0.061254  0.398931  0.533329  1.000000  0.698424  0.467917 -0.171274   
x5 -0.289361 -0.156304  0.496763  0.698424  1.000000  0.280129 -0.208277   
x6  0.198796  0.711134  0.032830  0.467917  0.280129  1.000000  0.416821   
x7  0.348698  0.413595 -0.139086 -0.171274 -0.208277  0.416821  1.000000   
x8  0.318677  0.834952 -0.258358  0.312757 -0.081234  0.701586  0.398868   

          x8  
x1  0.318677  
x2  0.834952  
x3 -0.258358  
x4  0.312757  
x5 -0.081234  
x6  0.701586  
x7  0.398868  
x8  1.000000  


（2）从R出发做主成分分析，求各主成分的贡献率及前两个主成分的累计贡献率

In [3]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# 标准化数据
scaler = StandardScaler()
data_scaled = scaler.fit_transform(data4_5.select_dtypes(include=[np.number]))
# 主成分分析
pca = PCA(n_components=8)
# 输出特征值
pca.fit(data_scaled)
eigenvalues = pca.explained_variance_
print("主成分分析结果：", eigenvalues)

# 计算各主成分贡献率
explained_variance_ratio = pca.explained_variance_ratio_
print("各主成分贡献率：", explained_variance_ratio)

# 前两个主成分累计贡献率
cumulative_variance = np.cumsum(explained_variance_ratio)
print("前两个主成分累计贡献率：", cumulative_variance[:2])

主成分分析结果： [3.20305685 2.44885145 0.95171105 0.73026761 0.51563    0.23781549
 0.13528528 0.05324435]
各主成分贡献率： [0.38703604 0.29590288 0.11499842 0.08824067 0.06230529 0.02873604
 0.01634697 0.00643369]
前两个主成分累计贡献率： [0.38703604 0.68293892]


(3)求出前两个主成分并解释其意义。按第一主成分得分将30个省、区、市排序，结果如何？

In [4]:
components = pca.components_[:2]
# 提取变量名
feature_names = ['x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8']
pc_df4_6 = pd.DataFrame(components, columns=feature_names, index=['PC1', 'PC2'])

print("前两个主成分对应的变量系数表：")
print(pc_df4_6.T)  

前两个主成分对应的变量系数表：
         PC1       PC2
x1  0.249607 -0.241238
x2  0.519234 -0.037607
x3 -0.018480  0.475439
x4  0.254092  0.538081
x5  0.021695  0.575449
x6  0.492663  0.134676
x7  0.317147 -0.260682
x8  0.509332 -0.087081


第一主成分PC1显著正贡献量是$ x_2、x_6、x_8 $，其次2是$ x_1、x_4、x_7 $，这表明人均副食支出、人均日用品支出、人均非商品支出高，反映出人均日常消费水平高，是花费在基本生存上的消费；

第二主成分显著贡献量是衣着、其他副食、烟酒茶、粮食和燃料，，第二主成分反映的是除去基本生存消费外的改善型消费。

In [5]:
# 按第一主成分对30个省份进行排序
data4_5['PC1'] = pca.transform(data_scaled)[:, 0]
sorted_data = data4_5.sort_values(by='PC1', ascending=False)

# 输出排序后的列表
print("按第一主成分排序的30个省份：")
print(sorted_data[['省(区、市)', 'PC1']])

按第一主成分排序的30个省份：
   省(区、市)       PC1
29     广东  7.013792
28     上海  3.303949
26     北京  1.822780
25     浙江  1.540971
22     海南  1.425115
20     福建  1.173626
21     广西  1.074576
23     天津  0.442873
24     江苏  0.155912
16     辽宁  0.045974
27     西藏 -0.135518
17     四川 -0.137193
18     山东 -0.143528
11     湖北 -0.173358
7      河北 -0.398903
9      宁夏 -0.437753
13     湖南 -0.526871
8      陕西 -0.623211
12     云南 -0.678096
10     新疆 -0.832496
6      青海 -1.132387
14     安徽 -1.134018
5      甘肃 -1.202439
1     内蒙古 -1.279703
15     贵州 -1.280871
2      吉林 -1.315817
3     黑龙江 -1.348335
4      河南 -1.511353
0      山西 -1.713280
19     江西 -1.994436


4.6


In [6]:
data4_6 = [
    [ 1,  60,  69,  62,  97,  69,  98],
    [ 2,  56,  53,  84, 103,  78, 107],
    [ 3,  80,  69,  76,  66,  99, 130],
    [ 4,  55,  80,  90,  80,  85, 114],
    [ 5,  62,  75,  68, 116, 130,  91],
    [ 6,  74,  64,  70, 109, 101, 103],
    [ 7,  64,  71,  66,  77, 102, 130],
    [ 8,  73,  70,  64, 115, 110, 109],
    [ 9,  68,  67,  75,  76,  85, 119],
    [10,  69,  82,  74,  72, 133, 127],
    [11,  60,  67,  61, 130, 134, 121],
    [12,  70,  74,  78, 150, 158, 100],
    [13,  66,  74,  78, 150, 131, 142],
    [14,  83,  70,  74,  99,  98, 105],
    [15,  68,  66,  90, 119,  85, 109],
    [16,  78,  63,  75, 164,  98, 138],
    [17, 103,  77,  77, 160, 117, 121],
    [18,  77,  68,  74, 144,  71, 153],
    [19,  66,  77,  68,  77,  82,  89],
    [20,  70,  70,  72, 114,  93, 122],
    [21,  75,  65,  71,  77,  70, 109],
    [22,  91,  74,  93, 118, 115, 150],
    [23,  66,  75,  73, 170, 147, 121],
    [24,  75,  82,  76, 153, 132, 115],
    [25,  74,  71,  66, 143, 105, 100],
    [26,  76,  70,  64, 114, 113, 129],
    [27,  74,  90,  86,  73, 106, 116],
    [28,  74,  77,  80, 116,  81,  77],
    [29,  67,  71,  69,  63,  87,  70],
    [30,  78,  75,  80, 105, 132,  80],
    [31,  64,  66,  71,  83,  94, 133],
    [32,  71,  80,  76,  81,  87,  86],
    [33,  63,  75,  73, 120,  89,  59],
    [34,  90, 103,  74, 107, 109, 101],
    [35,  60,  76,  61,  99, 111,  98],
    [36,  48,  77,  75, 113, 124,  97],
    [37,  66,  93,  97, 136, 112, 122],
    [38,  74,  70,  76, 109,  88, 105],
    [39,  60,  74,  71,  72,  90,  71],
    [40,  63,  75,  66, 130, 101,  90],
    [41,  66,  80,  86, 130, 117, 144],
    [42,  77,  67,  74,  83,  92, 107],
    [43,  70,  67, 100, 150, 142, 146],
    [44,  73,  76,  81, 119, 120, 119],
    [45,  78,  90,  77, 122, 155, 149],
    [46,  73,  68,  80, 102,  90, 122],
    [47,  72,  83,  68, 104,  69,  96],
    [48,  65,  60,  70, 119,  94,  89],
    [49,  52,  70,  76,  92,  94, 100],
]
df4_6 = pd.DataFrame(data4_6, columns=["编号","x1","x2","x3","y1","y2","y3"]).set_index("编号")

In [7]:
# 计算样本协方差矩阵S
S = df4_6.cov()

# 基于S做主成分分析
pca_s = PCA(n_components=6)
pca_s.fit(S)

# 计算各主成分贡献率
explained_variance_ratio_s = pca_s.explained_variance_ratio_
print("基于样本协方差S各主成分贡献率：", explained_variance_ratio_s.round(4))
feature_names = ['x1', 'x2', 'x3', 'y1', 'y2', 'y3']
pc_df4_6 = pd.DataFrame(pca_s.components_, columns=feature_names, index=['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6'])
print("基于样本协方差S前六个主成分对应的变量系数表：")
print(pc_df4_6.T)

基于样本协方差S各主成分贡献率： [0.6971 0.1732 0.1198 0.0073 0.0026 0.    ]
基于样本协方差S前六个主成分对应的变量系数表：
         PC1       PC2       PC3       PC4       PC5       PC6
x1  0.011793  0.055402 -0.108576  0.806422 -0.363969  0.449681
x2 -0.026452 -0.107650  0.150281  0.018865  0.755391  0.627818
x3  0.008537  0.086840 -0.027449 -0.586843 -0.494582  0.634536
y1  0.857909 -0.197948 -0.465652 -0.038118  0.078008  0.020954
y2  0.462137 -0.105408  0.864549  0.058630 -0.154916 -0.020917
y3  0.222509  0.963071  0.024435  0.007216  0.149049 -0.010891


基于协方差矩阵S的分析：保留两个主成分，PC1+PC2>85%

PC1主导变量为$ y_1、y_2、y_3 $，反映餐后血糖的绝对水平，尤其是餐后1小时血糖（$ y_1 $）的核心贡献。

PC2主导变量为$ y_3 $，捕捉餐后3小时血糖（$y_3 $）的独立变异，可能提示部分受试者血糖回落延迟。


In [8]:
# 相关系数矩阵R
R = df4_6.corr()

# 基于R做主成分分析
pca_r = PCA(n_components=6)
pca_r.fit(R)

# 计算各主成分贡献率
explained_variance_ratio_r = pca_r.explained_variance_ratio_
print("基于相关系数矩阵R各主成分贡献率：", explained_variance_ratio_r.round(4))

# 提取变量系数
pc_df4_6_r = pd.DataFrame(pca_r.components_, columns=feature_names, index=['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6'])
print("基于相关系数矩阵R的前六个主成分对应的变量系数表：")
print(pc_df4_6_r.T)

基于相关系数矩阵R各主成分贡献率： [0.3641 0.2879 0.2157 0.0963 0.0359 0.    ]
基于相关系数矩阵R的前六个主成分对应的变量系数表：
         PC1       PC2       PC3       PC4       PC5       PC6
x1  0.105426 -0.371571  0.716611 -0.080552  0.409412  0.403959
x2  0.674094  0.288123 -0.076371  0.171245 -0.311738  0.574670
x3 -0.003909 -0.409932 -0.663238 -0.354558  0.322753  0.402709
y1 -0.521365  0.357067  0.157198 -0.527732 -0.340808  0.425816
y2 -0.197848  0.616923 -0.102832  0.298058  0.668823  0.183098
y3 -0.472761 -0.320601 -0.073931  0.686393 -0.259919  0.359935


基于相关系数矩阵R的分析：保留三个主成分

PC1主导变量$ x_2、y_1、y_3 $，反映空腹$ x_2 $与餐后血糖（$ y_1、y_3 $）呈负相关，可能反映基础代谢率与餐后血糖响应的平衡。

PC2主导变量$ y_2、x_1、x_3 $，反映餐后2小时血糖（$ y_2 $）与空腹血糖的反向关系，体现胰岛素调节效率。

PC3主导变量$ x_1、x_3 $,反映空腹不同时间点（$ x_1 $与$ x_3 $）的血糖波动，可能反映测量时间或个体稳定性差异。

对于本题优先选择相关系数矩阵R：①空腹和餐后血糖具有不同生理意义，需避免量纲差异带来的偏差；②更全面揭示血糖调节机制（如胰岛素响应、空腹稳定性）。

4.9
从样本协方差矩阵 $\Sigma$ 和样本相关系数矩阵R出发做典型相关分析，求各典型变量及典型相关系数，检验各典型变量对是否显著相关 ($\alpha = 0.05$)。比较两种情况下的结果的异同。

In [9]:
from sklearn.cross_decomposition import CCA
from scipy.stats import chi2

# 原始数据
data4_9 = {
    "x1": [
        191, 195, 181, 183, 176, 208, 189, 197, 188, 192,
        179, 183, 174, 190, 188, 163, 195, 186, 181, 175,
        192, 174, 176, 197, 190
    ],
    "x2": [
        155, 149, 148, 153, 144, 157, 150, 159, 152, 150,
        158, 147, 150, 159, 151, 137, 155, 153, 145, 140,
        154, 143, 139, 167, 163
    ],
    "y1": [
        179, 201, 185, 188, 171, 192, 190, 189, 197, 187,
        186, 174, 185, 195, 187, 161, 183, 173, 182, 165,
        185, 178, 176, 200, 187
    ],
    "y2": [
        145, 152, 149, 149, 142, 152, 149, 152, 159, 151,
        148, 147, 152, 157, 158, 130, 158, 148, 146, 137,
        152, 147, 143, 158, 150
    ]
}

# 构造 DataFrame 并拆分 X, Y
df = pd.DataFrame(data4_9)
X = df[['x1', 'x2']].values
Y = df[['y1', 'y2']].values
n = X.shape[0]
p, q = X.shape[1], Y.shape[1]

def cca_analysis(X, Y, scale):
    """
    对 X, Y 做典型相关分析
    scale=False 时基于样本协方差矩阵
    scale=True 时基于样本相关系数矩阵
    """
    cca = CCA(n_components=min(p, q), scale=scale)
    cca.fit(X, Y)
    U, V = cca.transform(X, Y)
    # 计算各典型变量对之间的相关系数
    canonical_corrs = [np.corrcoef(U[:, i], V[:, i])[0, 1] 
for i in range(min(p, q))]
    # Wilks' Lambda 检验
    results = []
    for i in range(len(canonical_corrs)):
        # 从 i 到最后一对的 (1 - rho^2) 连乘积
        lambda_val = np.prod([1 - canonical_corrs[j]**2 
for j in range(i, len(canonical_corrs))])
        df_chi = (p - i) * (q - i)
        stat = - (n - 1 - (p + q + 1) / 2) * np.log(lambda_val)
        p_value = chi2.sf(stat, df_chi)
        results.append({
            'canonical_pair': i + 1,
            'corr': canonical_corrs[i],
            'Wilks_lambda': lambda_val,
            'chi2_stat': stat,
            'df': df_chi,
            'p_value': p_value
        })
    return cca, canonical_corrs, results

# 基于协方差矩阵的 CCA
cca_cov, r_cov, res_cov = cca_analysis(X, Y, scale=False)
# 基于相关系数矩阵的 CCA
cca_cor, r_cor, res_cor = cca_analysis(X, Y, scale=True)

print("----- 基于协方差矩阵的 CCA 结果 -----")
print("典型相关系数：", np.round(r_cov, 6))
print("X 端典型系数矩阵：\n", np.round(cca_cov.x_weights_, 6))
print("Y 端典型系数矩阵：\n", np.round(cca_cov.y_weights_, 6))
print("Wilks' Lambda 检验：")
for item in res_cov:
    print(item)

print("\n----- 基于相关系数矩阵的 CCA 结果 -----")
print("典型相关系数：", np.round(r_cor, 6))
print("X 端典型系数矩阵：\n", np.round(cca_cor.x_weights_, 6))
print("Y 端典型系数矩阵：\n", np.round(cca_cor.y_weights_, 6))
print("Wilks' Lambda 检验：")
for item in res_cor:
    print(item)


----- 基于协方差矩阵的 CCA 结果 -----
典型相关系数： [0.788508 0.05374 ]
X 端典型系数矩阵：
 [[ 0.624539  0.780993]
 [ 0.780993 -0.624539]]
Y 端典型系数矩阵：
 [[ 0.530786  0.847506]
 [ 0.847506 -0.530786]]
Wilks' Lambda 检验：
{'canonical_pair': 1, 'corr': np.float64(0.7885079162945556), 'Wilks_lambda': np.float64(0.37716288143970733), 'chi2_stat': np.float64(20.96417997947144), 'df': 4, 'p_value': np.float64(0.00032188971838857856)}
{'canonical_pair': 2, 'corr': np.float64(0.05373970442428279), 'Wilks_lambda': np.float64(0.9971120441683907), 'chi2_stat': np.float64(0.062180881478969025), 'df': 1, 'p_value': np.float64(0.8030815915559357)}

----- 基于相关系数矩阵的 CCA 结果 -----
典型相关系数： [0.788508 0.05374 ]
X 端典型系数矩阵：
 [[ 0.726998 -0.68664 ]
 [ 0.68664   0.726998]]
Y 端典型系数矩阵：
 [[ 0.6838  -0.72967]
 [ 0.72967  0.6838 ]]
Wilks' Lambda 检验：
{'canonical_pair': 1, 'corr': np.float64(0.7885079162945554), 'Wilks_lambda': np.float64(0.37716288143970766), 'chi2_stat': np.float64(20.964179979471417), 'df': 4, 'p_value': np.float64(0.00032188

典型相关系数：无论是基于样本协方差矩阵，还是基于样本相关系数矩阵做CCA，计算得到的两对典型相关系数均为$ \rho_1 = 0.7885, \rho_2 = 0.5537 $。这表明第一对典型变量之间的相关程度约为 0.7885，第二对几乎不相关。

典型变量系数：虽然两种方法下得到的系数数值不同，但它们对应的典型变量在原始尺度上的组合含义是相同的——两者都能提取出与第一对典型相关最强的线性组合；只不过当用 $R$ 时，数据先做了均值归一化，所以权重向量的方向和比例与直接用 $\Sigma$ 时不同。

显著性检验：

1.样本容量 $n=25$，$p=q=2$，对于第 $i$ 对典型相关，Wilks’ λ 计算及检验统计量如下：

第一对（$i=1$）：

$\Lambda_1 \approx 0.3771629$，

$\chi^2$ 统计量 $\approx 20.9642$， $\mathrm{df}=(2-0+0)(2-0+0)=4$，

$p \approx 0.00032$。
因此拒绝“第一对典型相关为 0” 的原假设：第一对典型相关显著不为 0。

第二对（$i=2$）：

$\Lambda_2 \approx 0.9971120$，

$\chi^2$ 统计量 $\approx 0.06218$， $\mathrm{df}=(2-1)(2-1)=1$，

$p \approx 0.80308$。
因此无法拒绝“第二对典型相关为 0” 的原假设：第二对典型变量之间不显著相关。

$\quad$上述 Wilks' λ 测试结果在两种做法下完全一致—：仅第一对典型变量显著，第二对不显著。

2.两种情况下的异同点：基于 $\Sigma$ 得到的权重体现的是“原始单位下”对典型变量的线性组合；基于 $R$ 得到的权重体现的是“标准化单位下”对典型变量的线性组合；但它们提取出的典型变量对在说明两组变量的最大相关方向上是一致的。